spec/lib/v1_alpha_fetcher_spec.rb

```diff
@@ -0,0 +1,127 @@
+
+RSpec.describe V1AlphaFetcher do
+  describe "#fetch" do
+    describe "success" do
+      before do
+        apis_response = {
+          "api-version": "api.gov.uk/v1alpha",
+          apis: [
+            "api-version": "api.gov.uk/v1alpha",
+            data: {
+              name: "test API",
+              description: "test description",
+              url: "www.foo.bar",
+              "documentation-url": "www.docs.foo.bar",
+              organisation: "Test Org",
+              contact: "Test Contact",
+            },
+          ],
+        }
+
+        stub_request(:get, "http://localhost/subpath/apis")
+          .with(headers: { "correlation-id" => /[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-4[a-fA-F0-9]{3}-[89abAB][a-fA-F0-9]{3}-[a-fA-F0-9]{12}/, "accept" => "application/vnd.uk.gov.api.v1alpha+json" })
+          .to_return(status: 200, body: apis_response.to_json)
+      end
+
+      it "parses APIs" do
+        v1_alpha_fetcher = described_class.new
+        api_catalogue = v1_alpha_fetcher.fetch "http://localhost/subpath"
+
+        expect(api_catalogue.organisations_apis.size).to eq(1)
+
+        _, org_apis = api_catalogue.organisations_apis.detect do |org, _|
+          org.name.casecmp?("Test Org")
+        end
+
+        test_api = org_apis.detect do |api|
+          api.name.casecmp?("test API")
+        end
+
+        expect(test_api).to have_attributes(
+          name: "test API",
+          description: "test description",
+          url: "www.foo.bar",
+          documentation: "www.docs.foo.bar",
+          provider: "test-org",
+          maintainer: "Test Contact",
+        )
+      end
+    end
+
+    describe "not accepted" do
+      it "raises NoVersionAcceptable" do
+        v1_alpha_fetcher = described_class.new
+        stub_request(:get, "http://localhost/subpath/apis")
+          .to_return(status: 406)
+
+        expect { v1_alpha_fetcher.fetch "http://localhost/subpath" }.to raise_error("Could not parse APIs from http://localhost/subpath/apis, no version supported")
+      end
+    end
+
+    describe "not found" do
+      it "raises NoVersionFound" do
+        v1_alpha_fetcher = described_class.new
+        stub_request(:get, "http://localhost/subpath/apis")
+          .to_return(status: 404)
+
+        expect { v1_alpha_fetcher.fetch "http://localhost/subpath" }.to raise_error("A client error caused retrieval of APIs from http://localhost/subpath/apis to fail")
+      end
+    end
+
+    describe "temporary error" do
+      it "raises TemporaryError" do
+        v1_alpha_fetcher = described_class.new
+        stub_request(:get, "http://localhost/subpath/apis")
+          .to_return(status: 500)
+
+        expect { v1_alpha_fetcher.fetch "http://localhost/subpath" }.to raise_error("A temporary error caused retrieval of APIs from http://localhost/subpath/apis to fail")
+      end
+    end
+
+    describe "bad request" do
+      it "raises ClientError" do
+        v1_alpha_fetcher = described_class.new
```

Comment:
we've duplicated this a lot, can we create it centrally?

Reply:
done